### PR TITLE
Support for registration of new EmitterGroup

### PR DESCRIPTION
**Pull request by davebenson**
See comments on https://github.com/lwes/lwes-java/pull/37

### CODE REVIEW

File: src/main/java/org/lwes/emitter/UDPEmitterGroupFactory.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
+package org.lwes.emitter;
```
Comment (pselden) [inline src/main/java/org/lwes/emitter/UDPEmitterGroupFactory.java]: Is this pretty much a rename of EmitterGroupBuilder as far as the contents are concerned?

File: src/main/java/org/lwes/emitter/EmitterGroupBuilder.java [hunk 2]
```diff
@@ -11,33 +11,47 @@
...
+   * @param factory An interface that spits out 
+   * EmitterGroups, give a Properties object and a prefix to look under.
+   */
+  public synchronized static void registerEmitterGroupFactory(String type, EmitterGroupFactory factory) {
+    //
+    // This pattern is, of course, not thread safe, but all registration
+    // should occur on the main java thread.
+    //
+    if (emitterGroupFactoryRegistry == null) {
+      emitterGroupFactoryRegistry = new HashMap<String, EmitterGroupFactory>();
```
Comment (pselden) [inline src/main/java/org/lwes/emitter/EmitterGroupBuilder.java]: Can you just add this as a property initializer?

File: src/main/java/org/lwes/emitter/EmitterGroupBuilder.java [hunk 3]
```diff
@@ -11,33 +11,47 @@
...
-  private static String STRATEGY_ALL = "all";
-  private static Pattern STRATEGY_M_OF_N = Pattern.compile("([\\d]*)ofN", Pattern.CASE_INSENSITIVE);
-  private static Pattern STRATEGY_NESTED = Pattern.compile("(([\\d|]*)ofN|all)_(([\\d]*)ofN|all)", Pattern.CASE_INSENSITIVE);
-  private static Pattern nestedTuplePattern = Pattern.compile("^((\\([^\\(]+?\\))(,)*)+$", Pattern.CASE_INSENSITIVE);
-  private static Pattern tuplePattern = Pattern.compile("\\([^\\(]+?\\)", Pattern.CASE_INSENSITIVE);
+
+  /**
+   * <p>Register a new 'type' of emitter.</p>
+   * <p>The default emitter is 'udp'.</p>
+   * <p>A 'gcs' emitter is available as a plugin.</p>
```
Comment (pselden) [inline src/main/java/org/lwes/emitter/EmitterGroupBuilder.java]: Where is it available?

File: src/main/java/org/lwes/emitter/EmitterGroupFilter.java [hunk 4]
```diff
@@ -39,4 +42,39 @@ public String toString() {
     for (String e : filtered) sb.append(e).append(" ");
     return sb.toString();
   }
+
+
+  // Helper method
+  public static EmitterGroupFilter fromProperties(Properties props, String prefix) {
```
Comment (pselden) [inline src/main/java/org/lwes/emitter/EmitterGroupFilter.java]: Is this exact copy/paste of getEmitterGroupFilter

File: src/main/java/org/lwes/emitter/EmitterGroupBuilder.java [hunk 5]
```diff
@@ -58,223 +72,19 @@
...
-
-  private static EmitterGroup buildNestedEmitterGroup(String prefix, String hostsStr, String strategyStr, int port, EmitterGroupFilter filter, double sampleRate, boolean emitHeartbeat, EventFactory factory) throws IOException {
-  String[] ratioConfig = strategyStr.split("_");
-  if (null == ratioConfig || ratioConfig.length != 2) {
-    throw new IllegalArgumentException("Invalid nested strategy config " + strategyStr);
-  }
-  if (!nestedTuplePattern.matcher(hostsStr).find()) {
-    throw new IllegalArgumentException("Invalid nested hosts config " + hostsStr);
+    String type = props.getProperty(prefix + "type");
+    if (type == null)
```
Comment (pselden) [inline src/main/java/org/lwes/emitter/EmitterGroupBuilder.java]: Style nit: can you add braces around the if statements here and below.

File: src/main/java/org/lwes/emitter/UDPEmitterGroupFactory.java (hunk 1 above)
Comment (davebenson) [reply src/main/java/org/lwes/emitter/UDPEmitterGroupFactory.java]: yes

File: src/main/java/org/lwes/emitter/EmitterGroupFilter.java (hunk 4 above)
Comment (davebenson) [reply src/main/java/org/lwes/emitter/EmitterGroupFilter.java]: Yes, it moved from EmitterGroup for cleanliness.

**Pull request merged**